contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,557 @@
+
+If a user wants to trigger the device plugin he only needs to request this
+through the same mechanism as OIRs in his Pod Spec.
+
+*Future*:
+
+In the future (beta) we would like to advertise each device and provide
+introspection capabilities (what pod uses what device).
+
+One solution would be to add a `Device` structure that could be defined
+as follows:
+
+```golang
+// E.g:
+// struct Device {
+//    ResourceName: "nvidia-gpu"
+//    ID: "GPU-fef8089b-4820-abfc-e83e-94318197576e"
+//    State: "Healthy",
+//    Properties: {
+//        "Family": "Pascal",
+//        "Memory": "4G",
+//        "ECC"   : "True",
+//    }
+//}
+type Device struct {
+	ResourceName string
+	ID           string
+	Status       DeviceHealthStatus
+	Properties   map[string]string
+}
+
+// Extend the node status
+type NodeStatus struct {
+	DevCapacity []Device
+}
+```
+
+We also introduce the `Devices` field in the Containers status so that user
+can know what devices were assigned to the container.
+
+```golang
+type ContainerStatus struct {
+	Devices []Device
+}
+```
+
+# Upgrading your cluster
+
+*TLDR:* 
+Given that we cannot guarantee that the Device Plugins are not running
+a daemon providing a critical service to Devices and when stopped will
+crash the running containers, it is up to the vendor to specify the
+upgrading scheme of their device plugin.
+
+However, If you are upgrading either Kubelet or any device plugin the safest way
+is to drain the node of all pods and upgrade.
+
+Depending on what you are upgrading and what changes happened then it
+is completely possible to only restart just Kubelet or just the device plugin.
+
+## Upgrading Kubelet
+
+This assumes that the Device Plugins running on the nodes fully implement the
+protocol and are able to recover from a Kubelet crash.
+
+Then, as long as the Device Plugin API does not change upgrading Kubelet can be done
+seamlessly through a Kubelet restart.
+
+*Currently:*
+As mentioned in the Versioning section, we currently expect the Device Plugin's
+API version to match exactly the Kubelet's Device Plugin API version.
+Therefore if the Device Plugin API version change then you will have to change
+the Device Plugin too.
+
+
+*Future:*
+When the Device Plugin API becomes a stable feature, versionning should be
+backward compatible and even if Kubelet has a different Device Plugin API,
+it should not require a Device Plugin upgrade.
+
+Refer to the versioning section for versioning scheme compatibility.
+
+## Upgrading Device Plugins
+
+Because we cannot enforce what the different Device Plugins will do, we cannot
+say for certain that upgrading a device plugin will not crash any containers
+on the node.
+
+It is therefore up to the Device Plugin vendors to specify if the Device Plugins
+can be upgraded without impacting any running containers.
+
+As mentioned earlier, the safest way is to drain the node before upgrading
+the Device Plugins.
+
```

Comment:
We may add some sentence here that we expect admins to set up monitoring services to watch for device plugin state. Exactly what kind of signals we would suggest to monitor may need more thinking. We can expand that part in the future.